### PR TITLE
Adding reader algorithm

### PR DESCRIPTION
**Pull request by jlingema**
Some changes in the DataHandle were necessary, some of which are temporary until we migrate to Gaudi v27r1 or later (see remarks in the code).

### CODE REVIEW

File: Examples/src/CheckCollection.cpp [hunk 1]
```diff
@@ -0,0 +1,51 @@
+#include "GaudiAlg/GaudiAlgorithm.h"
+
+#include "FWCore/DataHandle.h"
+
+#include "datamodel/MCParticleCollection.h"
+#include "datamodel/TrackHitCollection.h"
+#include "datamodel/GenVertex.h"
+
+
+class CheckCollection : public GaudiAlgorithm {
```
Comment (hegner) [inline Examples/src/CheckCollection.cpp]: CheckCollection sounds too generic to me. Can you use a name like FCCTestConsumer or something like that?

File: FWCore/FWCore/DataHandle.h [hunk 2]
```diff
@@ -61,9 +80,40 @@ DataHandle<T>::DataHandle() {}
...
+    if (LIKELY(m_isGoodType)) {
+      DataObjectHandle<DataWrapper<T> >::setRead();
+      std::cout << "returning static cast" << std::endl;
+      return static_cast<DataWrapper<T>*>(dataObjectp)->getData();
+    } else {
+      // The reader does not know the specific type of the collection. So we need a reinterpret_cast if the handle was
+      // created by the reader.
+      DataWrapper<podio::CollectionBase>* tmp = static_cast<DataWrapper<podio::CollectionBase>*>(dataObjectp);
+      DataObjectHandle<DataWrapper<T> >::setRead();
+      return reinterpret_cast<const T*>(tmp->collectionBase());
```
Comment (hegner) [inline FWCore/FWCore/DataHandle.h]: I am sure we can introduce some better type safety here. Let's discuss in person. Shouldn't be a blocker for this pull request though.

File: FWCore/components/PodioInput.h [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+// forward declarations
+// from FWCore:
+class PodioDataSvc;
+
+
+/** @class PodioInput FWCore/components/PodioInput.h PodioInput.h
+ *
+ *  Class that allows to read ROOT files written with PodioOutput
+ *  TODO: Automatically load needed libraries (needs changes to scheduler), check nEvents and stop run if EOF is reached
```
Comment (hegner) [inline FWCore/components/PodioInput.h]: make a JIRA ticket out of it ;-)

File: FWCore/components/PodioInput.h [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  /// ROOT file name the input is read from. Set by option filename
+  std::string m_filename;
+  /// Name of collections to read. Set by option collections (this is temporary)
+  std::vector<std::string> m_collectionNames;
+  /// Collection IDs (retrieved with CollectionIDTable from ROOT file, using collection names)
+  std::vector<int> m_collectionIDs;
+  /// Data service: needed to register objects and get collection IDs. Just an observing pointer.
+  PodioDataSvc* m_podioDataSvc;
+  /// Counter of the event number
+  int m_eventNum;
```
Comment (hegner) [inline FWCore/components/PodioInput.h]: why not put the counter into the service?
Once we go for multithreading that may be require anyhow.

File: FWCore/src/PodioDataSvc.cpp [hunk 5]
```diff
@@ -38,6 +39,13 @@ StatusCode PodioDataSvc::clearStore()    {
   return StatusCode::SUCCESS ;
 }
 
+void PodioDataSvc::setCollectionIDs(podio::CollectionIDTable* collectionIds) {
+  if (m_collectionIDs != nullptr) {
```
Comment (hegner) [inline FWCore/src/PodioDataSvc.cpp]: that makes me wonder whether a smart pointer could be something...

Comment (jlingema) [inline FWCore/src/PodioDataSvc.cpp]: As discussed: I agree but some changes to PODIO interfaces are necessary to make that a neat solution.

File: FWCore/components/PodioInput.h (hunk 4 above)
Comment (jlingema) [inline FWCore/components/PodioInput.h]: As discussed, I'll do this in a follow-up PR that moves the ownership of file, reader and event-store to the DataSvc.

File: FWCore/FWCore/DataHandle.h (hunk 2 above)
Comment (jlingema) [inline FWCore/FWCore/DataHandle.h]: I added `dynamic_casts` (done once) further up to check if it is safe to do the `static_cast` and `reinterpret_cast` here.

**Pull request merged**